### PR TITLE
Use `huggingface_hub` cache

### PR DESCRIPTION
**Pull request by lhoestq**
wip

- use `hf_hub_download()` from `huggingface_hub` for HF files
- `datasets` cache_dir is still used for:
  - caching datasets as Arrow files (that back `Dataset` objects)
  - extracted archives, uncompressed files
  - files downloaded via http (datasets with scripts)
- I removed code that were made for http files (and also the dummy_data / mock_download_manager stuff that happened to rely on them and have been legacy for a while now)

### CODE REVIEW

File: src/datasets/utils/file_utils.py [hunk 1]
```diff
@@ -190,32 +169,52 @@ def cached_path(
...
+                    force_download=download_config.force_download,
+                    proxies=download_config.proxies,
+                )
+            except (
+                huggingface_hub.utils.RepositoryNotFoundError,
+                huggingface_hub.utils.EntryNotFoundError,
+                huggingface_hub.utils.RevisionNotFoundError,
+                huggingface_hub.utils.GatedRepoError,
+            ) as e:
+                raise FileNotFoundError(str(e)) from e
```
Comment (lhoestq) [inline src/datasets/utils/file_utils.py:201]: that's the important part cc @Wauplin

File: src/datasets/utils/file_utils.py [hunk 2]
```diff
@@ -190,32 +169,52 @@ def cached_path(
 
     if is_remote_url(url_or_filename):
         # URL, so get it from the cache (downloading if necessary)
+        # HF HTTP URLs are prepared as to hf:// URLs
```
Comment (Wauplin) [inline src/datasets/utils/file_utils.py:172]: not sure how to interpret this comment

File: src/datasets/utils/file_utils.py [hunk 3]
```diff
@@ -190,32 +169,52 @@ def cached_path(
...
-            download_desc=download_config.download_desc,
-            disable_tqdm=download_config.disable_tqdm,
-        )
+        if url_or_filename.startswith("hf://"):
+            resolved_path = huggingface_hub.HfFileSystem(
+                endpoint=config.HF_ENDPOINT, token=download_config.token
+            ).resolve_path(url_or_filename)
+            try:
+                output_path = huggingface_hub.HfApi(
+                    endpoint=config.HF_ENDPOINT,
```
Comment (Wauplin) [inline src/datasets/utils/file_utils.py:182]:
```suggestion
```

This is already the default value in `huggingface_hub` (parsed from the same environment variable)

File: src/datasets/utils/file_utils.py [hunk 4]
```diff
@@ -190,32 +169,52 @@ def cached_path(
...
+                    token=download_config.token,
+                    library_name="datasets",
+                    library_version=__version__,
+                    user_agent=get_datasets_user_agent(download_config.user_agent),
+                ).hf_hub_download(
```
Comment (Wauplin) [inline src/datasets/utils/file_utils.py:187]: Top!

File: src/datasets/utils/file_utils.py [hunk 5]
```diff
@@ -190,32 +169,52 @@ def cached_path(
...
-            max_retries=download_config.max_retries,
-            token=download_config.token,
-            ignore_url_params=download_config.ignore_url_params,
-            storage_options=storage_options,
-            download_desc=download_config.download_desc,
-            disable_tqdm=download_config.disable_tqdm,
-        )
+        if url_or_filename.startswith("hf://"):
+            resolved_path = huggingface_hub.HfFileSystem(
+                endpoint=config.HF_ENDPOINT, token=download_config.token
```
Comment (Wauplin) [inline src/datasets/utils/file_utils.py:178]:
```suggestion
                token=download_config.token
```

This is already the default value in huggingface_hub (parsed from the same environment variable)

File: src/datasets/utils/file_utils.py [hunk 6]
```diff
@@ -1172,7 +913,7 @@ def _prepare_single_hop_path_and_storage_options(
         client_kwargs = storage_options.pop("client_kwargs", {})
         storage_options["client_kwargs"] = {"trust_env": True, **client_kwargs}  # Enable reading proxy env variables
         if "drive.google.com" in urlpath:
-            response = http_head(urlpath)
+            response = requests.head(urlpath, timeout=10)
```
Comment (Wauplin) [inline src/datasets/utils/file_utils.py:912]: Same comment as above about `requests.head` vs `get_session().head`. Maybe worth doing a pass on the `datasets` codebase in a separate PR

**Review by Wauplin — COMMENTED** [review]
Thanks for working on this @lhoestq! :tada: :tada: :tada: 

I did a first pass and left a few minor comments. Looks good!

File: src/datasets/utils/file_utils.py (hunk 3 above)
Comment (lhoestq) [reply src/datasets/utils/file_utils.py:182]: `datasets` users can modify `config. HF_ENDPOINT ` so I'd rather keep it

File: src/datasets/utils/file_utils.py (hunk 2 above)
Comment (lhoestq) [reply src/datasets/utils/file_utils.py:172]: I'll make it clearer. Basically even if someone passes a HTTP HF URL like `https://huggingface.co/datasets/.../data.csv` then it will be converted to `hf://datasets/.../data.csv`, that we can parse to get repo_id/revision/etc

File: src/datasets/utils/file_utils.py (hunk 6 above)
Comment (lhoestq) [reply src/datasets/utils/file_utils.py:912]: they are the only uses of `requests` in `datasets` left :) I'll use `get_session().head` though it's not a big deal imo

File: src/datasets/utils/file_utils.py (hunk 2 above)
Comment (Wauplin) [reply src/datasets/utils/file_utils.py:172]: ah I see, thanks :+1:

File: src/datasets/utils/file_utils.py (hunk 3 above)
Comment (Wauplin) [reply src/datasets/utils/file_utils.py:182]: Do you mean users monkey-patching a constant value at runtime without using the environment variable? I feel this is not something we should promote/support

Comment (lhoestq) [reply src/datasets/utils/file_utils.py:182]: `dataset-viewer` does it in its tests to switch between prod and testing endpoints :p

Comment (Wauplin) [reply src/datasets/utils/file_utils.py:184]: ok ok, maybe a topic for a separate PR then. It still feels wrong to me to handle endpoints in various places (both in `huggingface_hub` and in `datasets`)

File: src/datasets/utils/file_utils.py (hunk 6 above)
Comment (Wauplin) [reply src/datasets/utils/file_utils.py]: great to know!

**Comment by severo** [comment]
Nice

Capture d’écran 2024-08-19 à 15 25 00

**Review by Wauplin — APPROVED** [review]
Thanks @lhoestq! I thought the switch would be a much more complex process but happy to realize it's not! :smile: I re-reviewed the PR and it looks good to me -to the extent of my knowledgeable-. Better to have other pairs of eyes for this one :)

**Comment by lhoestq** [comment]
fyi the CI failure on test_py310_numpy2 is unrelated to this PR (it's a dependency install failure)

**Review by albertvillanova — COMMENTED** [review]
The CI error on test_py310_numpy2 has been temporarily fixed by:
- #7114

File: src/datasets/commands/dummy_data.py [hunk 7]
```diff
@@ -1,468 +0,0 @@
-import fnmatch
```
Comment (albertvillanova) [inline src/datasets/commands/dummy_data.py:1]: The removal of deprecated code has been addressed in a separate dedicated PR:
- #6996: https://github.com/huggingface/datasets/pull/6996/commits/5d4687e4ee1dba2b14e59117492d7f75ea66800a

**Review by albertvillanova — COMMENTED** [review]
The removal of deprecated code has been addressed in a separate dedicated PR:
-

**Review by albertvillanova — APPROVED** [review]
Thanks.

**Comment by julien-c** [comment]
yay! is this in a shipped release?

**Comment by lhoestq** [comment]
we can do one in the coming days once @albertvillanova is back

**Comment by albertvillanova** [comment]
We have made a release and this feature is now included.
